examples/generative/ddim.py

```diff
@@ -0,0 +1,839 @@
+residual connections help the information propagate through the network properly.
+* In the literature it is common to use
+[attention layers](https://keras.io/api/layers/attention_layers/multi_head_attention/)
+at lower resolutions for better global coherence. I omitted it for simplicity.
+* We disable the learnable center and scale parameters of the batch normalization layers,
+since the following convolution layers make them redundant.
+* We initialize the last convolution's kernel to all zeros as a good practice, making the
+network predict only zeros after initialization, which is the mean of its targets. This
+will improve behaviour at the start of training and make the mean squared error loss
+start at exactly 1.
+"""
+
+
+def get_network():
+    def sinusoidal_embedding(x):
+        embedding_min_frequency = 1.0
+        frequencies = tf.exp(
+            tf.linspace(
+                tf.math.log(embedding_min_frequency),
+                tf.math.log(embedding_max_frequency),
+                embedding_dims // 2,
+            )
+        )
+        angular_speeds = 2.0 * math.pi * frequencies
+        embeddings = tf.concat(
+            [tf.sin(angular_speeds * x), tf.cos(angular_speeds * x)],
+            axis=3,
+        )
+        return embeddings
+
+    def ResidualBlock(width):
+        def apply(x):
+            input_width = x.shape[3]
+            if input_width == width:
+                residual = x
+            else:
+                residual = layers.Conv2D(width, kernel_size=1)(x)
+            x = layers.BatchNormalization(center=False, scale=False)(x)
+            x = layers.Conv2D(
+                width, kernel_size=3, padding="same", activation=keras.activations.swish
+            )(x)
+            x = layers.Conv2D(width, kernel_size=3, padding="same")(x)
+            x = layers.Add()([x, residual])
+            return x
+
+        return apply
+
+    def DownBlock(width):
+        def apply(x):
+            x, skips = x
+            for _ in range(block_depth):
+                x = ResidualBlock(width)(x)
+                skips.append(x)
+            x = layers.AveragePooling2D(pool_size=2)(x)
+            return x
+
+        return apply
+
+    def UpBlock(width):
+        def apply(x):
+            x, skips = x
+            x = layers.UpSampling2D(size=2, interpolation="bilinear")(x)
+            for _ in range(block_depth):
+                x = layers.Concatenate()([x, skips.pop()])
+                x = ResidualBlock(width)(x)
+            return x
+
+        return apply
+
+    noisy_images = keras.Input(shape=(image_size, image_size, 3))
+    noise_variances = keras.Input(shape=(1, 1, 1))
+
+    e = layers.Lambda(sinusoidal_embedding)(noise_variances)
+    e = layers.UpSampling2D(size=image_size, interpolation="nearest")(e)
+
+    x = layers.Conv2D(widths[0], kernel_size=1)(noisy_images)
+    x = layers.Concatenate()([x, e])
+
+    skips = []
+    for width in widths[:-1]:
```

Comment:
Done, along with image_size and block_depth.

Reply:
Should embedding_dims and embedding_max_frequency be constructor arguments as well?

Reply:
I decided not to put the embedding hyperparams into the constructor, because they are only needed for the Lambda layer for which it is better if they are global.